galoy-client/src/client/mod.rs

```diff
@@ -0,0 +1,233 @@
+mod error;
+mod queries;
+
+use futures::stream::{self, Stream};
+use graphql_client::reqwest::post_graphql;
+use reqwest::{
+    header::{HeaderValue, AUTHORIZATION},
+    Client as ReqwestClient,
+};
+
+pub use error::*;
+pub use queries::*;
+
+pub struct LastTransactionCursor(String);
+impl From<String> for LastTransactionCursor {
+    fn from(string: String) -> Self {
+        Self(string)
+    }
+}
+
+impl LastTransactionCursor {
+    fn inner(self) -> String {
+        self.0
+    }
+}
+
+#[derive(Debug)]
+pub struct StablesatsWalletsBalances {
+    pub btc_wallet_balance: Option<queries::SignedAmount>,
+    pub usd_wallet_balance: Option<queries::SignedAmount>,
+}
+
+#[derive(Debug, Clone)]
+pub struct GaloyClient {
+    client: ReqwestClient,
+    config: GaloyClientConfig,
+    jwt: String,
+}
+
+#[derive(Debug, Clone)]
+pub struct GaloyClientConfig {
+    pub api: String,
+    pub code: String,
+    pub phone_number: String,
+}
+
+pub(crate) struct GaloyAuthToken {
+    pub inner: Option<String>,
+}
+
+impl GaloyClient {
+    pub async fn connect(config: GaloyClientConfig) -> Result<Self, GaloyClientError> {
+        let jwt = Self::login_jwt(config.clone()).await?;
+        let jwt = match jwt.inner {
+            Some(jwt) => jwt,
+            None => {
+                return Err(GaloyClientError::Authentication(
+                    "Empty authentication token".to_string(),
+                ))
+            }
+        };
+        let client = ReqwestClient::builder()
+            .default_headers(
+                std::iter::once((
+                    AUTHORIZATION,
+                    HeaderValue::from_str(&format!("Bearer {}", jwt)).unwrap(),
+                ))
+                .collect(),
+            )
+            .build()?;
+
+        Ok(Self {
+            client,
+            config,
+            jwt,
+        })
+    }
+
+    pub async fn authentication_code(
```

Comment:
When do we need this? To login we aways already have the code right?

Reply:
I think I should probably switch to using this library for galoy-cli?

in that case this would be useful for now

Reply:
I've thought about this and think they should remain separate.

- different use cases
- different lifecycles
- different teams

I don't see any upside for unifying atm. But we can discuss at a standup if you have a strong opinion but for me this is a classic case of DRY considered harmful. What is the trigger for unifying when its just going to create friction?

Reply:
I need in the cli things like:
- get/create address
- get list of transactions
- get balance
which has all been done here already

I could for sure copy the code or redo it, but the use case seems close enough to me that using a single library seems to make sense.

Things like batching that are specific to cli would not be in the library but on top of it.

Reply:
Its the `close enough` that I am worried about here. If I were in your position I would copy / paste.
But if you feel you'll be better served by depending on this sub-crate go ahead. Just be aware that there are likely to be breaking changes as we start using it in client code and know more about the needs.

In particular I could easily see the return types and arguments and initialisation to go through some iteration before were at a point that it stabilises for the use cases in this repo.

Not being able to iterate liberally (ie. needing to take external use cases into account and avoid breaking changes) causes coordination overhead that will slow everyone down.

Reply:
ok fair. I don't have to rush to use this library. I can do it in a couple of weeks once stablesats v2 is in prod.